pkg/util/mount/mount.go

```diff
@@ -84,6 +87,11 @@ type Interface interface {
 	// MakeDir creates a new directory.
 	// Will operate in the host mount namespace if kubelet is running in a container
 	MakeDir(pathname string) error
+	// KubeletPath returns path which kubelet can access whether it is running in a container or not
+	KubeletPath(pathname string) string
+	// GetRootPathAndMajorMinor returns root path relative to its mountpoint and
+	// major:minor of filesystem containing given path.
+	GetRootPathAndMajorMinor(pathname string) (string, string, error)
```

Comment:
Do we need this interface? Isn't the information in the MountInfo struct?

Reply:
@msau42 

It's possible to have same root paths in list of MountInfos. I wrote this interface to find right mount entry by comparing root path relative to its mountpoint and device major:minor in list of MountInfos. There are two cases, if they are self bind mounted, we don't need this interface. 

### If directories is not self bind mounted, this interface is needed

```
# two different filesystems 
mkdir -p /mnt/tmpfs{1,2}
mount -t tmpfs tmpfs /mnt/tmpfs1
mount -t tmpfs tmpfs /mnt/tmpfs2
# two directorys with same path in different filesystems
mkdir -p /mnt/tmpfs{1,2}/vol
# mount to another place to use
mkdir -p /mnt/dst{1,2}
mount --bind /mnt/tmpfs1/vol /mnt/dst1
mount --bind /mnt/tmpfs2/vol /mnt/dst2
#  grep -P '/mnt/dst(1|2)' /proc/self/mountinfo
62 24 0:84 /vol /mnt/dst1 rw,relatime shared:45 - tmpfs tmpfs rw
67 24 0:85 /vol /mnt/dst2 rw,relatime shared:46 - tmpfs tmpfs rw
#  grep -P '/mnt/tmpfs(1|2)' /proc/self/mountinfo
59 24 0:84 / /mnt/tmpfs1 rw,relatime shared:45 - tmpfs tmpfs rw
61 24 0:85 / /mnt/tmpfs2 rw,relatime shared:46 - tmpfs tmpfs rw
```

### If directories is self bind mounted, this interface is unnecessary
 
```
# two different filesystems 
mkdir -p /mnt/tmpfs{1,2}
mount -t tmpfs tmpfs /mnt/tmpfs1
mount -t tmpfs tmpfs /mnt/tmpfs2
# two directorys with same path in different filesystems
mkdir -p /mnt/tmpfs{1,2}/vol
# self-bind
mount --bind /mnt/tmpfs1/vol /mnt/tmpfs1/vol
mount --bind /mnt/tmpfs2/vol /mnt/tmpfs2/vol
# mount to another place to use
mkdir -p /mnt/dst{1,2}
mount --bind /mnt/tmpfs1/vol /mnt/dst1
mount --bind /mnt/tmpfs2/vol /mnt/dst2
# grep -P '/mnt/dst(1|2)' /proc/self/mountinfo
68 24 0:84 /vol /mnt/dst1 rw,relatime shared:45 - tmpfs tmpfs rw
69 24 0:85 /vol /mnt/dst2 rw,relatime shared:46 - tmpfs tmpfs rw
# grep -P '/mnt/tmpfs(1|2)' /proc/self/mountinfo
59 24 0:84 / /mnt/tmpfs1 rw,relatime shared:45 - tmpfs tmpfs rw
61 24 0:85 / /mnt/tmpfs2 rw,relatime shared:46 - tmpfs tmpfs rw
62 59 0:84 /vol /mnt/tmpfs1/vol rw,relatime shared:45 - tmpfs tmpfs rw
67 61 0:85 /vol /mnt/tmpfs2/vol rw,relatime shared:46 - tmpfs tmpfs rw
```

We can find root path ("/vol") and major:minor ("0:84") by path`/mnt/tmpfs1/vol` in mount entry 62. Should we assume directory volumes are always self bind mounted?

Reply:
Thanks for the explanation.  You are right, we cannot assume directory volumes are always self bind mounted.

I still have a question about why we need an interface method for this though.  `ListMountInfos` will return every entry in /proc/self/mountinfo, and it will contain the information about the mount root and device numbers.  Can you explain why in `MountGetRefsBySource`, we need to call this `GetRootPath` method when the `ListMountInfos` should already have all the information we need?  Instead of an interface, can `GetRootPath` just be a helper method?

Reply:
> Can you explain why in MountGetRefsBySource, we need to call this GetRootPath method when the ListMountInfos should already have all the information we need?

Because  in `/proc/self/mountinfo`, bind mount entry for `/mnt/dst1` does not use absolute path (`/mnt/tmpfs1/vol`) as mount source, it only contains its root path (`/vol`) and filesystem information (`0:84`), e.g.

```
62 24 0:84 /vol /mnt/dst1 rw,relatime shared:45 - tmpfs tmpfs rw
```

And same path can exist on more than one filesystems. `/mnt/tmpfs1/vol`, `/mnt/tmpfs2/vol` have same root path `/vol` under their filesystems. So for directory bind mounting, we need to use root path (path relative to its fs root) and filesystem (device `major:minor`) to uniquely identify bind source.

In case 1,  we cannot find its root path and filesystem in `/proc/self/mountinfo`, because no mount entries contains information for `/mnt/tmpfs1/vol`, because its actually a normal directory. So we need `GetRootPathAndMajorMinor()` to find out information we need by accessing its filesystem.

If case 2, if it's self bind mounted, there exists a mount entry 62, from which we known filesystem (`0:84`) and its path (`/vol`) is mounted at `/mnt/tmpfs1/vol`. So we can know `/mnt/tmpfs1/vol` is also mounted at `/mnt/dst1` by mount entry 68, because they have same filesystem (`0:84`) and mount root path (`/vol`).

> Instead of an interface, can GetRootPath just be a helper method?

OK. I'll change that.